# FIT5196 Task 1 in Assessment 1
#### Student Name: Girish D Bhatta
#### Student ID: 29270863

Date: 19/08/2018

Version: 1.0

Environment: Python 3.6 and Jupyter notebook

Libraries used: please include the main libraries you used in your assignment here, e.g.,:
* re (for regular expression, included in Anaconda Python 3.6) 
* json (for conversion of list of dictionaries to a json Object, included in Anaconda Python 3.6) 

## 1.  Import libraries 

In [3]:
# Code to import libraries. As part of this assessment task, 're' and 'json' libraries will be used for parsing the 
# input data and creating json and xml formats respectively.
import re
import json

## 2. Parse 29270863.dat File

In order to begin the parsing process, we have to perform superficial analysis or preprocessing to understand the kind of headings/keys are present in the Job posting data provided in the aforementioned file.

###     2.1. Reading the contents of the file.

In [4]:
#code to read the contents of the file
file_handler = open("29270863.dat","r")

#storing the content in a variable for further processing
input_text = file_handler.read()

#closing the file once the reading is done.
file_handler.close()

###  2.2 Splitting the contents of the file

On observation of "29270863.dat" file, it can be concluded that each job posting is separated by a string of '-' dashes. Therefore, splitting the contents of the file using '------------------------------' as a demiliter is a good way to begin with the parsing the data to required formats.

In [5]:
# splitting the contents by the string of '-'.
job_data = input_text.split('------------------------------')

# Since each job is separted by a the demiliter, the resultant list will have an empty string appended to the list.
# hence, we need to get rid of it.
del job_data[len(job_data)-1]

print("The file 29270863.dat has " + str(len(job_data)) + " job postings.")

The file 29270863.dat has 30942 job postings.


### 2.3 Analysing the keys/headings of the file.

Before we dive deep into the parsing of the file, we need analyse the keys/headings of the file. This is necessary in order to facilitate proper processing of data and to make sure that there is no data leakage.

<i>Upon close observation it can found out that every heading/tile of a job posting has a common pattern. Each heading <u> starts at the beginning of the line follwed by a ':', which essentially marks the end of the heading/title, and the data related to that field follows</u>. we capture these patterns in a regular expression.</i>

In [6]:
# The input_keys will have a list of all the keys/fields. The regular expression captures all keys. i.e, keys with all
# upper case characters, upper case characters beginning with a '_', upper case characters separated by '/' etc. similarly
# for all the lower case character and subsequent combinations as well.
input_keys = re.findall(r"(^_?[A-Z]+\s?[_/]*\s*[A-Z]*\s*[A-Z]*:|^_[a-z]+:|^[a-z]+_?[a-z]*:)",input_text,re.MULTILINE)
print("The file has a total of " + str(len(input_keys)) + " keys/headings") 

#printing the first 100 keys 
print(input_keys[0:100])

# code snippet to count the number of occurrences of each of the obtained keys
key_dict = {}
for each_key in input_keys:
    if each_key not in key_dict:
        key_dict[each_key] = 1
    else:
        key_dict[each_key] += 1

print("\n\nThe file has a total of " + str(len(key_dict)) + " distinct keys ")
key_dict

The file has a total of 323367 keys/headings
['ID:', 'PROCEDURE:', 'to:', 'LOCATIONS:', 'DEADLINES:', '_TTL:', 'ABOUT:', 'REQ_QUALS:', 'REMUNERATION/\nJOB RESPONSIBILITIES:', 'START DATE:', 'REMUNERATION:', '_description:', 'ID:', 'JOB_PROC:', '_LOC:', 'DEAD_LINE:', 'TITLES:', 'ABOUT:', 'QUALIFICATION:', 'REMUNERATION/\nJOB RESPONSIBILITIES:', 'START DATE:', 'remuneration:', 'DESCRIPTION:', 'ID:', 'PROCEDURES:', '_LOC:', 'deadline:', 'JOB_T:', 'COMPANYS_INFO:', 'REQ_QUALS:', 'JOB RESPONSIBILITIES:', 'START DATE:', 'salary:', 'DESCRIPTION:', 'ID:', 'PROCEDURE:', 'LOCATION:', 'deadline:', 'TITLES:', '_info:', 'REQUIRED QUALIFICATIONS:', 'REMUNERATION/\nRESP:', 'DATES:', 'SALARY:', 'DESCRIPTION:', 'JOB DESCRIPTION:', 'ID:', 'procedures:', 'LOCATION:', 'DEAD_LINE:', 'JOB_T:', 'REQUIRED QUALIFICATIONS:', 'JOB_RESPS:', 'START DATE:', 'JOB DESCRIPTION:', 'ID:', 'PROCEDURES:', '_LOC:', 'DEADLINES:', 'JOB TITLE:', 'about_company:', 'QUALIFS:', 'JOB RESPONSIBILITIES:', 'START DATE:', 'REMUNERATI

{'ABOUT COMPANY:': 11473,
 'ABOUT:': 4330,
 'APPLICATION_DEADL:': 3985,
 'APPLICATION_DL:': 4373,
 'COMPANYS_INFO:': 5810,
 'DATES:': 3863,
 'DATE_START:': 3284,
 'DEADLINES:': 5694,
 'DEAD_LINE:': 11564,
 'DESCRIPTION:': 5778,
 'EDUCATION/\nJOB RESPONSIBILITIES:': 1,
 'ID:': 30942,
 'JOB DESCRIPTION:': 11740,
 'JOB RESPONSIBILITIES:': 6178,
 'JOB TITLE:': 11662,
 'JOB_DESC:': 4053,
 'JOB_LOC:': 4121,
 'JOB_PROC:': 4295,
 'JOB_PROCS:': 4055,
 'JOB_RESPS:': 4350,
 'JOB_SAL:': 3971,
 'JOB_T:': 4016,
 'LOCATION:': 11390,
 'LOCATIONS:': 5741,
 'PROCEDURE:': 11545,
 'PROCEDURES:': 5655,
 'QUALIFICATION:': 5827,
 'QUALIFS:': 4107,
 'REMUNERATION/\nDATES:': 119,
 'REMUNERATION/\nDESCRIPTION:': 2,
 'REMUNERATION/\nJOB DESCRIPTION:': 3,
 'REMUNERATION/\nJOB RESPONSIBILITIES:': 5342,
 'REMUNERATION/\nREMUNERATION:': 16,
 'REMUNERATION/\nRESP:': 1834,
 'REMUNERATION/\nRESPONSIBILITY:': 2652,
 'REMUNERATION/\nSALARY:': 27,
 'REMUNERATION/\nSTART DATE:': 382,
 'REMUNERATION:': 5794,
 'REQUIRED QUAL

Although the regular expression seems to be working fine extracting all the required fields for us, but it also is capturing the data with similar pattern as part of the value. values like 'to:','address:','and:','are:' etc. Therefore we need to refine our regular expression further, which will only extract the required fields and not unnecessary data which appears as part of the value data. 

In [7]:
# the refined regular expression will take care of all the fileds that appear in lower case and have the same pattern as that of
# fields. we need to perform all the steps performed previosuly to check if all the necessary keys are getting captured.

res = re.findall(r"(^_?[A-Z]+\s?[_/]*\s*[A-Z]*\s*[A-Z]*:|^_[a-z]+:|^[a-z]+_[a-z]*:|^salary:|^titles?:|^locations?:|^remuneration:|^descriptions?:|^responsibility:|^responsibilities:|^qualifications?:|^procedures?:|^deadline:|^about:)"
                 ,input_text,re.MULTILINE)

print("The file has a total of " + str(len(res)) + " keys/headings") 

#printing the first 100 keys 
print(res[0:100])

# code snippet to count the number of occurrences of each of the obtained keys
key_dict = {}
for each_key in res:
    if each_key not in key_dict:
        key_dict[each_key] = 1
    else:
        key_dict[each_key] += 1

print("\n\nThe file has a total of " + str(len(key_dict)) + " distinct keys ")
key_dict

The file has a total of 319279 keys/headings
['ID:', 'PROCEDURE:', 'LOCATIONS:', 'DEADLINES:', '_TTL:', 'ABOUT:', 'REQ_QUALS:', 'REMUNERATION/\nJOB RESPONSIBILITIES:', 'START DATE:', 'REMUNERATION:', '_description:', 'ID:', 'JOB_PROC:', '_LOC:', 'DEAD_LINE:', 'TITLES:', 'ABOUT:', 'QUALIFICATION:', 'REMUNERATION/\nJOB RESPONSIBILITIES:', 'START DATE:', 'remuneration:', 'DESCRIPTION:', 'ID:', 'PROCEDURES:', '_LOC:', 'deadline:', 'JOB_T:', 'COMPANYS_INFO:', 'REQ_QUALS:', 'JOB RESPONSIBILITIES:', 'START DATE:', 'salary:', 'DESCRIPTION:', 'ID:', 'PROCEDURE:', 'LOCATION:', 'deadline:', 'TITLES:', '_info:', 'REQUIRED QUALIFICATIONS:', 'REMUNERATION/\nRESP:', 'DATES:', 'SALARY:', 'DESCRIPTION:', 'JOB DESCRIPTION:', 'ID:', 'procedures:', 'LOCATION:', 'DEAD_LINE:', 'JOB_T:', 'REQUIRED QUALIFICATIONS:', 'JOB_RESPS:', 'START DATE:', 'JOB DESCRIPTION:', 'ID:', 'PROCEDURES:', '_LOC:', 'DEADLINES:', 'JOB TITLE:', 'about_company:', 'QUALIFS:', 'JOB RESPONSIBILITIES:', 'START DATE:', 'REMUNERATION:', '

{'ABOUT COMPANY:': 11473,
 'ABOUT:': 4330,
 'APPLICATION_DEADL:': 3985,
 'APPLICATION_DL:': 4373,
 'COMPANYS_INFO:': 5810,
 'DATES:': 3863,
 'DATE_START:': 3284,
 'DEADLINES:': 5694,
 'DEAD_LINE:': 11564,
 'DESCRIPTION:': 5778,
 'EDUCATION/\nJOB RESPONSIBILITIES:': 1,
 'ID:': 30942,
 'JOB DESCRIPTION:': 11740,
 'JOB RESPONSIBILITIES:': 6178,
 'JOB TITLE:': 11662,
 'JOB_DESC:': 4053,
 'JOB_LOC:': 4121,
 'JOB_PROC:': 4295,
 'JOB_PROCS:': 4055,
 'JOB_RESPS:': 4350,
 'JOB_SAL:': 3971,
 'JOB_T:': 4016,
 'LOCATION:': 11390,
 'LOCATIONS:': 5741,
 'PROCEDURE:': 11545,
 'PROCEDURES:': 5655,
 'QUALIFICATION:': 5827,
 'QUALIFS:': 4107,
 'REMUNERATION/\nDATES:': 119,
 'REMUNERATION/\nDESCRIPTION:': 2,
 'REMUNERATION/\nJOB DESCRIPTION:': 3,
 'REMUNERATION/\nJOB RESPONSIBILITIES:': 5342,
 'REMUNERATION/\nREMUNERATION:': 16,
 'REMUNERATION/\nRESP:': 1834,
 'REMUNERATION/\nRESPONSIBILITY:': 2652,
 'REMUNERATION/\nSALARY:': 27,
 'REMUNERATION/\nSTART DATE:': 382,
 'REMUNERATION:': 5794,
 'REQUIRED QUAL

Now, we can think that the refinement has yielded a fair result cutting down the number of keys from 229 to 63 keys. Further, we need to standardise the names of the fields by considering the nmeanings they convey and replacing the variants of a key/field with a standard key/field value. For example 'ABOUT COMPANY','ABOUT','_info','about_company' all mean the same and hence standarization is necessary for further processing.

### 2.3 Dissection/Explanation of regex to extract keys/fields

WE now Dissect our regular expression to capture keys/Fields of each document.

The regular expression we have used is : 

<b>(^_?[A-Z]+\s?[_/]*\s*[A-Z]*\s*[A-Z]*:|^_[a-z]+:|^[a-z]+_[a-z]*:|^salary:|^titles?:|^locations?:|^remuneration:|^descriptions?:|^responsibility:|^responsibilities:|^qualifications?:|^procedures?:|^deadline:|^about:)</b>

The regular expression has 3 major sections:

1. <b>^_?[A-Z]+\s?[_/]*\s*[A-Z]*\s*[A-Z]*:</b> ----> This part of regular expression handles all the keys which are uppercase. We have seen keys like _TTL which begin with '_' hence we begine with ^_?. the ^ caret ensures that this pattern is captured at the beginning of the string. [A-Z]+\s?[_/]*\s*[A-Z]*\s*[A-Z]*: This part matches all the keys which are in uppercase followed by a space or a tab captured by \s. followed by another underscore '_' or '/' character which is captured by [_/]* this part. this can be followed by any number of uppercase characters which is captured by *\s*[A-Z]*:. It is important to notice that every regex to capture fields/keys should end with a colon ':' this will serve as a delimiter to capture keys/fields.
    
2. <b>^_[a-z]+:|^[a-z]+_[a-z]*:</b> ----> This part of regex matches all the keys which have lower case characters. these are of two types. The ones which like _loc which represent the location of the job posting start with an underscore. Hence  ^_[a-z]+: this regex matches all the keys which start with the patter of '_' followed by 1 or more lower case characters and ending with a colon. ^[a-z]+_[a-z]*: this regex checks for keys with two words joined by a '_' between them. this will capture keys like "job_t".
    
3. <b>^salary:|^titles?:|^locations?:|^remuneration:|^descriptions?:|^responsibility:|
    responsibilities:|^qualifications?:|^procedures?:|^deadline:|^about:</b> ----> This regex is used to eliminate indesirable tokens like "'to:','address:','and:','are:'". Although this regex looks quiet absurd, during our prelimiary analysis, we had observed that extraction of undesirable tokens occured only with lower case keys/fields. As we do know that the keys we need to extract will belong to the aforementioned fields, we just looks for them in the regex, thereby, eliminating unnecessary extraction of undesirable tokens like "to:","with:","address:" etc. The reason for using only these specific strings as a part of our regex can be attributed to the preliminary analysis of the number of occurences of these keys. we have used those keys which appear most of the times. 

### 2.4 Standardization of Keys and extraction of data

We now start to standardize all the keys of the of the split content, extract the data for the all the fields and persist the extracted data in a dictionary and append it to a list. that means, each dictioanry will represent each job posting and a list will have a collection of job postings.

In [8]:
# A list that will contain all the job postings,basically it will be a list of dictionaries, where each dictionary represents a
# job posting.
all_jobs = []

# This code snippet iterates over the job_data list, where each list item is a job posting, it creates dictionary for each list
# list item, extracts all the fields, creates a dynamic regex to extract all the values and then store it in the dictionary with
# appropriate key/field names and lastly append the dictionary to a list.
for job in job_data:
   count = 0 # to keep track of the keys/fileds of the job posting
   job = job.strip() # to get rid of extra white spaces in the text.
   each_job = {} #dictionary to capture each job posting's data
   res = re.findall(r"(^_?[A-Z]+\s?[_/]*\s*[A-Z]*\s*[A-Z]*:|^_[a-z]+:|^[a-z]+_[a-z]*:|^salary:|^titles?:|^locations?:|^remuneration:|^descriptions?:|^responsibility:|^responsibilities:|^qualifications?:|^procedures?:|^deadline:|^about:)"
                 ,job,re.MULTILINE) # regex to capture all the keys/fields of the job posting
    
   # loop to iterate over the list of keys,look for patterns, standardize the keys and extract the associated value 
   for i in range(len(res)):
        count += 1        
        if count == len(res): # if the the key is the last element of the list
            my_regex = res[i]+r"(.*)" #dyanmic regular expression that with a capturing group, 
                                    #that starts with the key and is not bounded by a ending key.
            
            # the keys are standardized in accordance to the keys mentioned in the sample.xml file   
            if re.search(r"(ID:)",res[i],re.IGNORECASE):
                each_job["_id"] = re.search(my_regex,job,re.DOTALL).group(1)
            elif re.search(r"(.*loc.*)",res[i],re.IGNORECASE):
                each_job["location"] = re.search(my_regex,job,re.DOTALL).group(1)
            elif re.search(r"(.*desc.*)",res[i],re.IGNORECASE):
                each_job["job_descriptions"] = re.search(my_regex,job,re.DOTALL).group(1)
            elif re.search(r"(.*resp.*)",res[i],re.IGNORECASE):
                each_job["job_responsibilities"] = re.search(my_regex,job,re.DOTALL).group(1)
            elif re.search(r"(.*qual.*)",res[i],re.IGNORECASE):
                each_job["required_qualifications"] = re.search(my_regex,job,re.DOTALL).group(1)
            elif re.search(r"(.*salary.*|.*remuneration.*|job_sal)",res[i],re.IGNORECASE):
                each_job["salary"] = re.search(my_regex,job,re.DOTALL).group(1)
            elif re.search(r"(.*proc.*)",res[i],re.IGNORECASE):
                each_job["application_procedure"] = re.search(my_regex,job,re.DOTALL).group(1)
            elif re.search(r"(.*dead.*|application_dl)",res[i],re.IGNORECASE):
                each_job["application_deadline"] = re.search(my_regex,job,re.DOTALL).group(1)
            elif re.search(r"(.*title.*|.*ttl.*|job_t)",res[i],re.IGNORECASE):
                each_job["title"] = re.search(my_regex,job,re.DOTALL).group(1)
            elif re.search(r"(.*about.*|.*info.*)",res[i],re.IGNORECASE):
                each_job["about_company"] = re.search(my_regex,job,re.DOTALL).group(1)
            elif re.search(r"(.*start.*|.*date.*)",res[i],re.IGNORECASE):
                each_job["start_date"] = re.search(my_regex,job,re.DOTALL).group(1)
        else:    
            my_regex = res[i]+r"(.*?)"+res[i+1] #dynamic regex that starts with a given key pointed by res[i], followed by
                                                # a capturing group, to capture the data of that field, until it reaches the key
                                                # pointed by res[i+1]
            
            if re.search(r"(ID:)",res[i],re.IGNORECASE):
                each_job["_id"] = re.search(my_regex,job,re.DOTALL).group(1)
            elif re.search(r"(.*loc.*)",res[i],re.IGNORECASE):
                each_job["location"] = re.search(my_regex,job,re.DOTALL).group(1)
            elif re.search(r"(.*desc.*)",res[i],re.IGNORECASE):
                each_job["job_descriptions"] = re.search(my_regex,job,re.DOTALL).group(1)
            elif re.search(r"(.*resp.*)",res[i],re.IGNORECASE):
                each_job["job_responsibilities"] = re.search(my_regex,job,re.DOTALL).group(1)
            elif re.search(r"(.*qual.*)",res[i],re.IGNORECASE):
                each_job["required_qualifications"] = re.search(my_regex,job,re.DOTALL).group(1)
            elif re.search(r"(.*salary.*|.*remuneration.*|job_sal)",res[i],re.IGNORECASE):
                each_job["salary"] = re.search(my_regex,job,re.DOTALL).group(1)
            elif re.search(r"(.*proc.*)",res[i],re.IGNORECASE):
                each_job["application_procedure"] = re.search(my_regex,job,re.DOTALL).group(1)
            elif re.search(r"(.*dead.*|application_dl)",res[i],re.IGNORECASE):
                each_job["application_deadline"] = re.search(my_regex,job,re.DOTALL).group(1)
            elif re.search(r"(.*title.*|.*ttl.*|job_t)",res[i],re.IGNORECASE):
                each_job["title"] = re.search(my_regex,job,re.DOTALL).group(1)
            elif re.search(r"(.*about.*|.*info.*)",res[i],re.IGNORECASE):
                each_job["about_company"] = re.search(my_regex,job,re.DOTALL).group(1)
            elif re.search(r"(.*start.*|.*date.*)",res[i],re.IGNORECASE):
                each_job["start_date"] = re.search(my_regex,job,re.DOTALL).group(1)
   all_jobs.append(each_job)

print("The total number of jobs are " + str(len(all_jobs)))

# printing the first 2 jobs
print(all_jobs[0:2]) 

The total number of jobs are 30942
[{'_id': ' 94582\n', 'application_procedure': ' To apply for this position, please submit a\ndetailed resume/CV and a cover letter in Armenian and English languages\nto: Haypost CJSC HR Department (6th floor, 22 Saryan Str., 0002 Yerevan,\nRepublic of Armenia) or e-mail those to: HRManager@.... \nPlease, mention in subject line of your e-mail the position title. \nOnly short listed candidates will be notified for the interview.\nPlease clearly mention in your application letter that you learned of\nthis job opportunity through Career Center and mention the URL of its\nwebsite - www.careercenter.am, Thanks.\n', 'location': ' Yerevan, Armenia\n', 'application_deadline': ' 20 May 2008\n', 'title': ' Marketing Specialist\n', 'about_company': '\n ""Orange Fitness and Tennis Club"" is a fitness club\noffering its customers a bundle of services directed to their health\nimprovement and active leisure.\n', 'required_qualifications': '\n - University degree in

There are certain keys like 'job_responsibilities','job_descriptions','required_qualifications' etc which have bullet points. these bullet points need to converted to list and and assigned back to a dictionary whose key will be the actual field names. This is necessary from the point of view of conversion to json and xml formats.

In [9]:
# This code snippet converts the bullet points, which begin with a '-' and intermediately terminated by ';' and permanently
# terminated by '.' into a list from a textual format. it looks for patterns in the values, and if such a pattern does exist, it 
# performs the necessary operation of splitting them by ';' as a delimiter and assigning it back to a dictionary whose key is 
# the field name itself.

for each_obj in all_jobs:
    for k in each_obj: 
        each_obj[k] = each_obj[k].strip()
        if re.search(r'^-.*;$|^-.*\.$',each_obj[k],re.DOTALL|re.MULTILINE): #looks for a pattern which has a '-' at the beginning
                                                                            #and either ends with a ';' or '.'
            obj = {} #dictionary to hold the data in key value pair, where value would be a list and key woould be the field name
                     #from the parent dictionary.
            if k == "job_responsibilities":
                obj["responsibility"] = each_obj[k].replace('-','').strip().split(';') #replacing the '-', stripping the whitespaces and splitting by
                                                                                       #delimiter ';'
                each_obj[k] = obj
            elif k == "job_descriptions":
                obj["description"] = each_obj[k].replace('-','').strip().split(';')
                each_obj[k] = obj
            elif k == "required_qualifications":
                obj["qualification"] = each_obj[k].replace('-','').strip().split(';')
                each_obj[k] = obj

#Printing the first 2 job postings
print(all_jobs[0:2])

[{'_id': '94582', 'application_procedure': 'To apply for this position, please submit a\ndetailed resume/CV and a cover letter in Armenian and English languages\nto: Haypost CJSC HR Department (6th floor, 22 Saryan Str., 0002 Yerevan,\nRepublic of Armenia) or e-mail those to: HRManager@.... \nPlease, mention in subject line of your e-mail the position title. \nOnly short listed candidates will be notified for the interview.\nPlease clearly mention in your application letter that you learned of\nthis job opportunity through Career Center and mention the URL of its\nwebsite - www.careercenter.am, Thanks.', 'location': 'Yerevan, Armenia', 'application_deadline': '20 May 2008', 'title': 'Marketing Specialist', 'about_company': '""Orange Fitness and Tennis Club"" is a fitness club\noffering its customers a bundle of services directed to their health\nimprovement and active leisure.', 'required_qualifications': {'qualification': ['University degree in Law or Economics', '\n At least 3 years 

<b>Post standardization and extraction, we can analyse the number of occurences of the keys/fields, which would give insights into the presence/absence of these fields across job postings.</b>

In [10]:
# this code snippet collects all the keys/fields of all the job postings and calculates the number of occurences of each of these
# keys. this would provide insights into the presences/absence of these fields across job postings

keys_list = []
for each_job in all_jobs:
    for each_key in each_job:
        keys_list.append(each_key)

key_dict = {}
for each_key in keys_list:
    if each_key not in key_dict:
        key_dict[each_key] = 1
    else:
        key_dict[each_key] += 1
 
print("All the distinct key and the number of their occurrences are :")
for k,v in key_dict.items():
    print(k," : ",v)
    


All the distinct key and the number of their occurrences are :
_id  :  30942
application_procedure  :  28845
location  :  28789
application_deadline  :  28715
title  :  28834
about_company  :  28777
required_qualifications  :  28839
job_responsibilities  :  28722
start_date  :  28275
salary  :  28926
job_descriptions  :  28821


We can observe that except the key "id" none of the keys appear accross all the job postings. On an average we can see that out of 30942 job postings all other keys are missing in 1500 files. We can therfore conclude that not all keys/fields appear in all job postings except the key "_id".

### 2.5 Dumping the extracted data to a json file

In this step we have to dump the extracted data to a json file. This can be done by making use of the 'dump' method by json library that was imported earlier.


In [14]:
# This code snippet adds a few more addons to the existing all_jobs data in terms of a dictionary that encapsualtes all the data 
# and places in the appropriate format as required.

all_job_obj = {} #the dictionary that will hold the entire extracted data
inner_obj = {} #dictionary, with a key to hold the all_job list data.
inner_obj["listing"] = all_jobs #assignement of extracted data to a "listing" key of inner_obj
all_job_obj["listings"] = inner_obj # assigning back to the all_job_obj dictionary, with "listings" as the key

# writing the contents of the all_job_obj to 29270863.json
fh = open("29270863.json","w")
fh.write(json.dumps(all_job_obj))
fh.close()

print("The json content was written to 29270863.json file.")


The json content was written to 29270863.json file.


### 2.6 Transforming the extracted Data to XML

Now that we have transformed the raw text data to json, we have to do the same to convert the 'all_jobs' list to an xml format and write to a file. since there are no functions available to do this task for us, we have to write the logic to convert the list of dictionaries to xml string and write to a file.

In [15]:
#This code snippet basically iterates over all_jobs list, extracts information and puts it in an XML format.


xml_str = "<listings>" #initializing a string variable with the beginning tag of the target XML document
for each_job in all_jobs: #iterating through all the jobs 
    for k in each_job: #checking for ID key
        if k == "_id":
            sub_xml = "<listing id = "+'"'+each_job[k]+'"'">" # if so, then adding the ID in listing tag in required style.
            xml_str += sub_xml
        elif k == "job_descriptions" or k == "job_responsibilities" or k == "required_qualifications": #checking for keys which 
                                                                    #have dictionaries as values.
            if type(each_job[k]) == dict:  #if true, an inner_xml string is created that concatenates the list items surrounded
                                           # by appropriate tags and is in turn concatenated with the global string xml_str
                inner_xml = ""
                for key in each_job[k]:
                    for each_item in each_job[k][key]:
                        inner_xml += "<"+key+">"+each_item+"</"+key+">"
                sub_xml = "<"+k+">"+inner_xml+"</"+k+">"
                xml_str += sub_xml
            else:
                sub_xml = "<"+k+">"+each_job[k]+"</"+k+">"
                xml_str += sub_xml
        else:
            sub_xml = "<"+k+">"+each_job[k]+"</"+k+">" # For other keys, the xml string is dynamically created and the data associated
                                                        # is fetched accordingly and concatenated in the right place.
            xml_str += sub_xml
    xml_str += "</listing>" #the tag that marks the end of a job posting is terminated with </listing> tag
xml_str += "</listings>" #once all the job postings are iterated over, the </listings> tag is appended to the end of the xml_str

#writing the xml string to 29270863.xml file 
fh = open("29270863.xml","w")
fh.write(xml_str)
fh.close()

print("The xml content was written to 29270863.xml file.")

The xml content was written to 29270863.xml file.


## 3. Summary

##### Section 1. Demonstrated the importing of essential libraries like re and json to carry out Data parsing.

##### Section 2. Showcased the Data parsing steps as follows:
<ul>
    <li><b>Section 2.1</b> This section deals with reading the contents of 29270863.dat file.</li>
    <li><b>Section 2.2</b> This section deals with splitting the contents of the file by a delimiter. thereby attaining the contents of each file as a list item. This step is necessary to carry out further processing.</li>
    <li><b>Section 2.3</b> This section explains in detail the way we arrived at our final regular expression which was used to extract keys/fields of each job posting</li>
    <li><b>Section 2.4</b> This section handles the standardization of all the keys and capturing data of respective keys and persisting them in a dictionary where each dictionary represents a job posting and each dictionary will have a set of keys and values, where keys represent the keys/fields of the job posting and value represents its corresponding data. This eventually will we stored in one single list</li>
    <li><b>Section 2.5</b> We perform certain post standardization analysis of the standardized keys/fields and check the number of files these occur in. That is followed by converting the list of dictionaries of job posting to appropriate format and then dumping it using the json package and eventually wrirting it to a 29270863.json file</li>
    <li><b>Section 2.6</b> Lastly we use the List of dictionaries to create an XML string and the same is subsequently written to 29270863.xml file</li>

</ul>


## 4. References

#### <li>https://docs.python.org/3/library/re.html</li>


#### <li>https://docs.python.org/3/library/json.html</li>
